File: 08.two-pointers/solve_cpp/leetcode_31.cpp
/*
 * Next Permutation
 */
#include <iostream>
#include <algorithm>

using namespace std;

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        // 1. pivot
        int pivot = nums.size() - 2;

        while (pivot >= 0) {
            if (nums[pivot] < nums[pivot + 1]) 
                break;
            pivot--;
        }

        if (pivot == -1) {
            sort(nums.begin(), nums.end());
        }

        // 2. swap point
        int j = pivot + 1;
        int swap_point = j;
        int min_value = nums[j];

        while (j < nums.size()) {
            if ((nums[pivot] < nums[j]) && (nums[j] < min_value)) {
                swap_point = j;
                min_value = nums[j];
            }
            j += 1;
        }

        cout << pivot << endl;
        cout << swap_point << endl;

        // 3. swap
        if (pivot > -1) {
            int tmp = nums[pivot];
            nums[pivot] = nums[swap_point];
            nums[swap_point] = tmp;
        }

        // 4. pivot+1 ~ asc 
        sort(nums.begin() + pivot + 1, nums.end());
    }

    void printVector(vector<int>& nums) {
        for (int num : nums) {
            cout << num << " ";
        }
        cout << endl;
    }
};

int main() {
    Solution solution;

    vector<int> v = {1,3,2,4};
    cout << "input : ";
    solution.printVector(v);

    solution.nextPermutation(v);
    cout << "solution : ";
    solution.printVector(v);

    return 0;
}

File: 08.two-pointers/solve_py/leetcode_31.py
def next_permutation(nums: list[int]):
    """
    정수 순열 배열의 사전적 순서
    1) 오른쪽에서 왼쪽으로 조회하면서 arr[i] < arr[i+1] 을 만족하는 i 찾기
    2) i+1 부터의 원소 중에서 arr[i] 보다 큰 값중 가장 작은 값의 인덱스를 j
    3) swap(i, j)
    4) i+1 부터의 배열을 오름차순 정리
    """
    # 1. pivot
    pivot = len(nums) - 2

    while pivot >= 0:
        if nums[pivot] < nums[pivot + 1]:
            break
        pivot -= 1

    if pivot == -1:
        sorted(nums)
        # return sorted(nums)

    # 2. swap_point
    j = pivot + 1
    swap_point = j
    min_value = nums[j]
    while j < len(nums):
        if (nums[pivot] < nums[j]) and (nums[j] < min_value):
            swap_point = j
            min_value = nums[j]
        j += 1

    # 3. swap
    if pivot > -1:
        tmp_val = nums[pivot]
        nums[pivot] = nums[swap_point]
        nums[swap_point] = tmp_val

    # 4. pivot + 1 부터 오름차순 정렬
    nums[pivot+1:] = sorted(nums[pivot+1:])

    print(nums)

    # return nums
